pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,487 @@
+	if err != nil {
+		if errors.IsNotFound(err) {
+			c.consistentHashRWMutex.Lock()
+			defer c.consistentHashRWMutex.Unlock()
+			delete(c.consistentHashMap, eipName)
+			return nil
+		}
+		return err
+	}
+
+	// updateConsistentHash refresh the consistentHashMap.
+	updateConsistentHash := func(eip *v1alpha2.ExternalIPPool) error {
+		nodeSel, err := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+		if err != nil {
+			return err
+		}
+		nodes, err := c.nodeLister.List(nodeSel)
+		if err != nil {
+			return err
+		}
+		aliveNodes := c.aliveNodes()
+		// Node alive and Node labels matches with ExternalIPPool nodeSelector.
+		var aliveAndMatchedNodes []string
+		for _, node := range nodes {
+			nodeName := node.Name
+			if aliveNodes.Has(nodeName) {
+				aliveAndMatchedNodes = append(aliveAndMatchedNodes, nodeName)
+			}
+		}
+		consistentHashMap := newNodeConsistentHashMap()
+		consistentHashMap.Add(aliveAndMatchedNodes...)
+		c.consistentHashRWMutex.Lock()
+		defer c.consistentHashRWMutex.Unlock()
+		c.consistentHashMap[eip.Name] = consistentHashMap
+		c.notify(eip.Name)
+		return nil
+	}
+
+	if err := updateConsistentHash(eip); err != nil {
+		return err
+	}
+	return nil
+}
+
+func newNodeConsistentHashMap() *consistenthash.Map {
+	return consistenthash.New(defaultVirtualNodeReplicas, defaultHashFn)
+}
+
+func (c *Cluster) handleClusterNodeEvents(nodeEvent *memberlist.NodeEvent) {
+	node, event := nodeEvent.Node, nodeEvent.Event
+	switch event {
+	case memberlist.NodeJoin, memberlist.NodeLeave:
+		// When a Node joins cluster, all matched ExternalIPPools consistentHash should be updated.
+		// when a Node leave cluster, the Node may failed or deleted,
+		// if Node has been deleted, affected ExternalIPPool should enqueue, deleteNode handler has processed
+		// if the Node failed, ExternalIPPools consistentHash maybe changed, affected ExternalIPPool should enqueue.
+		coreNode, err := c.nodeLister.Get(node.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// Node has been deleted, deleteNode handler has processed.
+				klog.ErrorS(err, "Processing Node event, not found", "eventType", event)
+				return
+			}
+			klog.ErrorS(err, "Processing Node event, get Node failed", "eventType", event)
+			return
+		}
+		affectedEIPs := c.filterEIPsFromNodeLabels(coreNode)
+		c.enqueueExternalIPPools(affectedEIPs)
+		klog.InfoS("Processed Node event", "eventType", mapNodeEventType[event], "nodeName", node.Name, "affectedExternalIPPoolNum", len(affectedEIPs))
+	default:
+		klog.InfoS("Processed Node event", "eventType", mapNodeEventType[event], "nodeName", node.Name)
+	}
+}
+
+// aliveNodes returns the list of nodeNames in the cluster.
+func (c *Cluster) aliveNodes() sets.String {
+	nodes := sets.NewString()
+	for _, node := range c.mList.Members() {
+		nodes.Insert(node.Name)
+	}
+	return nodes
+}
+
+// ShouldSelectEgress the local Node in the cluster holds the same consistent hash ring for each ExternalIPPool,
```

Comment:
Probably you missed a verb after ShouldSelectEgress

Reply:
```
-// ShouldSelectEgress the local Node in the cluster holds the same consistent hash ring for each ExternalIPPool,
-// when selecting an owner Node for an Egress, the local Node labels must match an ExternalIPPool nodeSelectors.
+// ShouldSelectEgress returns true if the local Node selected as the owner Node of the Egress,
+// the local Node in the cluster holds the same consistent hash ring for each ExternalIPPool,
```